src/_guides/language/concurrency/index.md

```diff
@@ -0,0 +1,403 @@
+[`readAsString()`]: https://api.dart.dev/stable/dart-io/File/readAsString.html
+
+Why does it matter whether a method is synchronous or asynchronous?
+It matters because most apps need to do more than one thing at a time.
+For example, an app might start an HTTP request,
+but need to update its display or respond to user input
+before the HTTP request completes.
+Asynchronous code helps apps stay responsive.
+
+### The async-await syntax
+The `async` and `await` keywords provide a declarative way to define asynchronous functions and use their results.
+
+Here’s an example of some synchronous code that blocks while waiting for file I/O:
+
+```dart
+void main() {
+  // Read some data.
+  final fileData = _readFileSync();
+  final jsonData = jsonDecode(fileData);
+
+  // Use that data.
+  print('Number of JSON keys: ${jsonData.length}');
+}
+
+String _readFileSync() {
+  final file = File(filename);
+  return file.readAsStringSync();
+}
+```
+
+Here’s similar code, but with changes (highlighted) to make it asynchronous:
+
+{% prettify dart tag=pre+code %}
+[!Future<void>!] main() [!async!] {
+  // Read some data.
+  final fileData = [!await!] _readFileAsync();
+  final jsonData = jsonDecode(fileData);
+
+  // Use that data.
+  print('Number of JSON keys: ${jsonData.length}');
+}
+
+[!Future<String>!] _readFileAsync() [!async!] {
+  final file = File(filename);
+  return [!await!] file.[!readAsString()!];
+}
+{% endprettify %}
+
+The `main()` function uses the `await` keyword in front of `_readFileAsync()`
+to let other Dart code (such as event handlers) use the CPU
+while native code (file I/O) executes.
+Using `await` also has the effect of
+converting the `Future<String>` returned by `_readFileAsync()` into a `String`.
+As a result, the `fileData` variable has the implicit type `String`.
+
+{{site.alert.note}}
+  The `await` keyword works only in functions that
+  have `async` before the function body.
+{{site.alert.end}}
+
+As the following figure shows,
+the Dart code pauses while `readAsString()` executes non-Dart code,
+in either the Dart virtual machine (VM) or the operating system (OS).
+Once `readAsString()` returns a value, Dart code execution resumes.
+
+![Flowchart-like figure showing app code executing from start to exit, waiting for native I/O in between](/guides/language/concurrency/images/basics-await.png)
+
+If you’d like to learn more about using `async`, `await`, and futures,
+visit the [asynchronous programming codelab][].
+
+[asynchronous programming codelab]: /codelabs/async-await
+
+
+## How isolates work
+
+Most modern devices have multi-core CPUs.
+To take advantage of all those cores,
+developers sometimes use shared-memory threads running concurrently.
+However, shared-state concurrency is
+[error prone](https://en.wikipedia.org/wiki/Deadlock) and
+can lead to complicated code.
+
+Instead of threads, all Dart code runs inside of isolates.
```

Comment:
It would be good to describe this in terms of existing OS concepts. 

For example, are isolates in the same process? Or different processes? Or is that a per-architecture implementation decision? Do they communicate via IPC, or shared memory, or something else?

Reply:
I don't know the answers to all of these questions. Maybe in v2?

Reply:
@mit-mit Is there someone in Aarhus who could draft a couple of sentences here while we're sleeping? :)

Reply:
I discussed this a bit, but the VM folks are hesitant to be too explicit as the VM use of processes and threads is considered a implementation detail, and subject to change. I'm still waiting for confirmation, but I think something like this might be acceptable?

```
Within an app, all Dart code runs in an isolate, conceptually similar to a process. Each Dart isolate has a single thread of execution and shares no memory with other isolates. Communication between isolates happens via message passing.
```

Reply:
I would **love it** if we could say something like that explicitly ("the VM use of processes and threads is considered a implementation detail, and subject to change"). Because it's hard to take the line that Dart doesn't use threads ("instead of threads") -- since of course the unit of execution in any modern OS is indeed a thread.

Reply:
> shares no memory with other isolates

This might be misleading. Isolates do share immutable data(strings, constants, etc) and code with isolates in the same isolate group: the point of lightweight isolates is that they share the same VM heap. Isolates do not share _mutable_ data (as in - changes to the data done by one isolate is not visible to another isolate).

Reply:
OK, so then, this should work? 

> Within an app, all Dart code runs in an isolate, conceptually similar to a process. Each Dart isolate has a single thread of execution and shares no mutable data with other isolates. Communication between isolates happens via message passing. Naturally, Dart's isolate model is built with underlying primitives provided by the operating system, but the Dart VM's use of processes and threads is considered an implementation detail.

Reply:
I discussed with Slava, and he's fine with this text:

> Within an app, all Dart code runs in an isolate. Each Dart isolate has a single thread of execution and shares no mutable data with other isolates. Communication between isolates happens via message passing. Naturally, Dart's isolate model is built with underlying primitives provided by the operating system, but the Dart VM's use of processes and threads is considered an implementation detail.

This leaves out the "conceptually similar to a process" which he found potentially very misleading: they are like processes in some ways but very different in others, and processes are typically very expensive, where isolates are light.

Reply:
How about this (_italicized changes_ are to clarify and to avoid via, naturally, some passive voice, and other writing anathemas):

> Within an app, all Dart code runs in an isolate. Each Dart isolate has a single thread of execution and shares no mutable objects with other isolates. _Instead of using shared memory, isolates communicate by passing messages._ Dart’s isolate model is built with underlying primitives, such as processes and threads, that the operating system provides; the Dart VM’s use of these primitives is an implementation detail _that this page doesn’t discuss._

Reply:
I've pushed this change and restaged it:

https://kw-dartlang-3.web.app/guides/language/concurrency

Reply:
After further discussion out of band, I'll delete "Instead of using shared memory."

Reply:
New change is up and restaged

Reply:
https://github.com/dart-lang/site-www/pull/3696/commits/09b735f2aa773e26c074572cf32212959666e732 LGTM, resolving this thread!